tests/trio108.py

```diff
@@ -554,6 +555,7 @@ async def foo_cm_1():
 
 @contextlib.asynccontextmanager
 async def foo_cm_2():
+    # Just imagine some fancy control flow so that only one of these executes.
     yield
     yield
     yield
```

Comment:
Oh no, wait - I might have completely misunderstood how asynccontextmanagers work? Is `__aenter__` only called on first time the function is called, and `__aexit__` on the last time that it is? Or is `__aenter__` before every time function execution is continued?
Cause if it's the former, this PR is bad (at least the asynccontextmanager part) and should be slightly modified to only be marking this as safe
```python
@asynccontextmanager
async def foo():
  yield
```
(which without the acm would have two errors)

Reply:
(All the fancy control flow logic for conditional execution of yields is implemented :wink: )

Reply:
and e.g.
```python
@asynccontextmanager
async def foo():
  while True:
    yield
```
isn't an error with this PR

Reply:
Those methods are called exactly once each, when you enter and exit the with-block respectively (which may be well after the function call!).

I don't think it's worth linting for too-many-yields though, because there's more-precise runtime error detection.

Reply:
Isn't that kind of the whole point of like 90% of 108 though? :P If they always just lead to false positives I'd disable them everywhere, but when the checks are written I don't see the point in explicitly disabling them for this one type of decorator. It's a trivial change to turn back on the too-many-yields check

Reply:
We're walking a fairly delicate balance here: it's not feasible to be precisely correct in all circumstances (for example, runtime codegen is a thing in Python!), so we're making tradeoffs between our false-alarm and missed-alarm rates.

IIRC, our current too-many-yields check is *pessimistic* about user code, and warns whenever there might be a problem - no missed alarms, but this leads to many false alarms on context-manager code.  I also just don't think that the multiple-yield thing is ~ever going to arise in practice in a context manager.

What we *could* do is just have the TRIO107 check active for async context managers; no matter what else is going on they should always checkpoint at least once before exiting (unless an exception is raised, etc.).

Reply:
[starts writing long message]

*oh*, I once again realized the error I'm making. I was associating `yield` too hard with iterators, and not realizing how context managers actually work (I guess I've never written my own context manager) and didn't properly grasp your statements about context managers only supposing to `yield` once.

Lemme read up on how to actually write a `contextmanager` and see where I land with what if anything 107/108 should do with them. There's also a possibility of creating a separate error code for yielding twice in a `contextmanager` (async or not) since the logic is already here, though that's not conceptually a great fit for a trio-specific plugin (but could always be default off).
But now that I've invested the time into learning how to do configuration flags, it's relatively easy to tune the strictness of checks with those. And for checks such as this it could make a lot of sense to CI with reasonable flags, but once in a blue moon run a false-alarm prone checker against the code to see if it warns against anything new. Another target for this could have been whether to be pessimistic about whether `asynccontextmanagers` checkpoints both on entry and exit. 

Thanks for your patience :heart: